syntax = "proto3";

package inventoryPackage;

service InventoryService {
  rpc AddPart(AddPartRequest) returns (AddPartResponse);
  rpc CheckPartsAvailability(CheckPartsRequest) returns (CheckPartsResponse);
  rpc ReserveParts(ReservePartsRequest) returns (ReservePartsResponse);
  rpc GetRequiredParts(GetRequiredPartsRequest) returns (GetRequiredPartsResponse);
  rpc ManageStockLevels(ManageStockLevelsRequest) returns (ManageStockLevelsResponse);
}

message PartInfo {
  string partId = 1;
  string partName = 2;
  int32 quantity = 3;
}

message CheckPartsRequest {
  repeated string partIds = 1;
}

message CheckPartsResponse {
  repeated PartInfo partsInfo = 1;
}

message ReservePartsRequest {
  string reservationId = 1;
  map<string, int32> partsToReserve = 2;  // a map for parts to reserve; part_id -> quantity
}

message ReservePartsResponse {
  bool success = 1;
  string message = 2;
}

message GetRequiredPartsRequest {
  string repairTypeName = 1;
}

message GetRequiredPartsResponse {
  repeated PartInfo partsInfo = 1;
}

message ManageStockLevelsRequest {
  map<string, int32> partsToUpdate = 1; // part_id -> quantity adjustment (positive for adding, negative for removing)
}

message ManageStockLevelsResponse {
  bool success = 1;
  string message = 2;
}

message AddPartRequest {
  PartInfo partInfo = 1;
}

message AddPartResponse {
  PartInfo partInfo = 1;
}

// message AddRepairTypeRequest {
//   RepairType repair_type = 1;
// }

// message AddRepairTypeResponse {
//   bool success = 1;
// }